dask/dataframe/tests/test_csv.py

```diff
@@ -0,0 +1,190 @@
+    blocks = []
+    for k in sorted(files):
+        b = files[k]
+        lines = b.split(b'\n')
+        blocks.append([b'\n'.join(bs) for bs in partition_all(2, lines)])
+
+    df = read_csv_from_bytes(blocks, header, expected.head(), {})
+    eq(df.compute().reset_index(drop=True),
+       expected.reset_index(drop=True), check_dtype=False)
+
+    expected2 = expected[['name', 'id']]
+    df = read_csv_from_bytes(blocks, header, expected2.head(),
+                             {'usecols': ['name', 'id']})
+    eq(df.compute().reset_index(drop=True),
+       expected2.reset_index(drop=True), check_dtype=False)
+
+
+def test_enforce_dtypes():
+    blocks = [[b'aa,bb\n1,1.0\n2.2.0', b'10,20\n30,40'],
+              [b'aa,bb\n1,1.0\n2.2.0', b'10,20\n30,40']]
+    head = pd.read_csv(BytesIO(blocks[0][0]), header=0)
+    dfs = read_csv_from_bytes(blocks, b'aa,bb\n', head, {}, enforce_dtypes=True,
+            collection=False)
+    dfs = compute(*dfs)
+    assert all(df.dtypes.to_dict() == head.dtypes.to_dict() for df in dfs)
+
+
+def test_read_csv_files():
+    with filetexts(files, mode='b'):
+        df = read_csv('2014-01-*.csv')
+        eq(df, expected, check_dtype=False)
+
+        fn = '2014-01-01.csv'
+        df = read_csv(fn)
+        expected2 = pd.read_csv(BytesIO(files[fn]))
+        eq(df, expected2, check_dtype=False)
+
+
+from dask.bytes.compression import compress, files as cfiles, seekable_files
+fmt_bs = [(fmt, None) for fmt in cfiles] + [(fmt, 10) for fmt in seekable_files]
+
+@pytest.mark.parametrize('fmt,blocksize', fmt_bs)
+def test_read_csv_compression(fmt, blocksize):
+    files2 = valmap(compress[fmt], files)
+    with filetexts(files2, mode='b'):
+        df = read_csv('2014-01-*.csv', compression=fmt, blocksize=blocksize)
+        eq(df.compute(get=get_sync).reset_index(drop=True),
+           expected.reset_index(drop=True), check_dtype=False)
+
+
+def test_warn_non_seekable_files(capsys):
+    files2 = valmap(compress['gzip'], files)
+    with filetexts(files2, mode='b'):
+        df = read_csv('2014-01-*.csv', compression='gzip')
+        out, err = capsys.readouterr()
+        assert 'gzip' in out
+        assert 'blocksize=None' in out
+
+        df = read_csv('2014-01-*.csv', compression='gzip', blocksize=None)
+        out, err = capsys.readouterr()
+        assert not out
+
+        with pytest.raises(NotImplementedError):
+            df = read_csv('2014-01-*.csv', compression='foo')
+
+
+def test_windows_line_terminator():
+    text = 'a,b\r\n1,2\r\n2,3\r\n3,4\r\n4,5\r\n5,6\r\n6,7'
+    with filetext(text) as fn:
+        df = read_csv(fn, blocksize=5, lineterminator='\r\n')
+        assert df.b.sum().compute() == 2 + 3 + 4 + 5 + 6 + 7
+        assert df.a.sum().compute() == 1 + 2 + 3 + 4 + 5 + 6
+
+
+def test_late_dtypes():
+    text = 'a,b\n1,2\n2,3\n3,4\n4,5\n5.5,6\n6,7.5'
+    with filetext(text) as fn:
+        df = read_csv(fn, blocksize=5, sample=10)
+        try:
+            df.b.sum().compute()
```

Comment:
This fails because the dtypes in the first block are int, but then we try to read floats?